_posts/2019-04-09-spring-security-in-action.md

```diff
@@ -0,0 +1,224 @@
+        response.setStatus(HttpStatus.UNAUTHORIZED.value());
+    }
+}
+```
+
+### VI. UserDetailService
+UserDetailService is the core interface which loads user-specific data. We mast realize loadUserByUsername() function to locate user and user's role.
+
+### VII. Rest APIs
+Once all authoritioin configure has been finished, you can enjoy your developing. To control your role and access you can simply add @PreAuthorize("hasAuthority('ADMIN_USER')") in your rest api declaration.
+```java
+@RequestMapping(value = "/users", method = RequestMethod.GET)
+@PreAuthorize("hasAuthority('ADMIN_USER')") //user with ADMIN_USER role have this access.
+public ResponseEntity<List<User>> getUsers() {
+    return new ResponseEntity<>(userService.findAllUsers(), HttpStatus.OK);
+}
+```
+
+## **Demo**
+
+### I. Apply for a token
+Client end can either use postman or curl to get a token.
+
+```sh
+curl client-id:client-password@localhost:8080/oauth/token -d grant_type=password -d username=admin.admin -d password=Test!123
+```
+
+You will get:
+```json
+{"access_token":"eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOlsic3ByaW5nLXNlY3VyaXR5LWRlbW8tcmVzb3VyY2UtaWQiXSwidXNlcl9uYW1lIjoiYWRtaW4uYWRtaW4iLCJzY29wZSI6WyJyZWFkIiwid3JpdGUiXSwiZXhwIjoxNTU0ODQ0NTQxLCJhdXRob3JpdGllcyI6WyJTVEFOREFSRF9VU0VSIiwiQURNSU5fVVNFUiJdLCJqdGkiOiI4MTM3Y2Q4OS0wMWMyLTRkMTgtYjA4YS05MjNkOTcxYjNhYzQiLCJjbGllbnRfaWQiOiJjbGllbnQtaWQifQ.1t_4xVT8xaAtisHaNT_nMRBLKfpiI0SZQ2bbEGxu6mk","token_type":"bearer","expires_in":43199,"scope":"read write","jti":"8137cd89-01c2-4d18-b08a-923d971b3ac4"}
+```
+
+### II. Use the token in role control
+Then use the token above to post a request which need authorition.
+```sh
+curl http://localhost:8080/demo/users -H "Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOlsic3ByaW5nLXNlY3VyaXR5LWRlbW8tcmVzb3VyY2UtaWQiXSwidXNlcl9uYW1lIjoiYWRtaW4uYWRtaW4iLCJzY29wZSI6WyJyZWFkIiwid3JpdGUiXSwiZXhwIjoxNTU0ODQ0NTQxLCJhdXRob3JpdGllcyI6WyJTVEFOREFSRF9VU0VSIiwiQURNSU5fVVNFUiJdLCJqdGkiOiI4MTM3Y2Q4OS0wMWMyLTRkMTgtYjA4YS05MjNkOTcxYjNhYzQiLCJjbGllbnRfaWQiOiJjbGllbnQtaWQifQ.1t_4xVT8xaAtisHaNT_nMRBLKfpiI0SZQ2bbEGxu6mk"
+```
+It will return:
+```json
+[{"id":1,"username":"jakob.he","firstName":"Jakob","lastName":"He","roles":[{"id":1,"roleName":"STANDARD_USER","description":"Standard User"}]},{"id":2,"username":"admin.admin","firstName":"Admin","lastName":"Admin","roles":[{"id":1,"roleName":"STANDARD_USER","description":"Standard User"},{"id":2,"roleName":"ADMIN_USER","description":"Admin User"}]}]
+```
+
+### III. Verify the token by JWT
+Put your token and signing key in [jwt.io](https://jwt.io), you will get following result.
+
+> ![jwt-verification.jpg](integration-blog/assets/2019-04-09-spring-security-in-action/jwt-verification.jpg)
+
+
+Let’s quickly go over, we have introduced what is Spring-Security, why we need to use it, and after all we have also realize a complete Spring-Security application including token management, token distribution, and rest APIs that requiring web authorization. 
```

Comment:
Let's quickly go over `what we have done: ` we have introduced what is Spring-Security and why we need to use it. We have also `implemented` a complete Spring-Security application `that included` token management, token distribution, and `REST APIs that are required for web authorization`.